### ISSUE TITLE
IdentityServerBuilder returns null after injecting the service in Startup's configure method

### ISSUE THREAD

**Issue by SpiritBob** [IdentityServer/IdentityServer4#4071]
**We can only help you if you are on the latest version. Either use the latest 2.x or 3.x**

Please only use the issue tracker for bug reports and/or feature requests. For general security questions, or free or commercial support options do __not__ use the issue tracker and instead see [here](http://docs.identityserver.io/en/latest/intro/support.html) for more details.

For bug reports,  include the relevant log files related to your issue. See here how to enable [logging](https://identityserver4.readthedocs.io/en/release/topics/logging.html). Delete this line once you have.

Finally, please keep the issue concise and to the point. If you paste in more code than the text for the issue you are reporting then we will most likely not read it. 

### Issue / Steps to reproduce the problem
Make a project in ASP.NET Core 3.1/3.0 (tested on 3.1)
Due to how Dependency injection is done in Startup.cs now, we can't add services to the Startup directly, but rather have to do it inside its Configure method.

Add ordinary configuration options to your identity server like so:
```
            services.AddIdentityServer()
                .AddInMemoryClients(ConfigureIdentityServer.GetClients())
                .AddInMemoryIdentityResources(ConfigureIdentityServer.GetIdentityResources())
                .AddInMemoryApiResources(ConfigureIdentityServer.GetApiResources())
                .AddProfileService<UserProfileService>();
```

In Startup's configure method, inject the service, like so:

`IIdentityServerBuilder identityBuilder = app.ApplicationServices.GetService<IIdentityServerBuilder>();`

This results in a null object returned. I've tried adding the service by using `services.AddIdentityServerBuilder();` but it yielded the same results.

**Comment by leastprivilege** [comment]
`IIdentityServerBuilder` is a helper to populate the DI container. You use it in `ConfigureServices` not `Configure`.

**Comment by SpiritBob** [comment]
@leastprivilege what if we want to implement our own population extension methods, and they need additional services, which can only be added inside Configure method?

**Comment by leastprivilege** [comment]
Configure is for building the pipeline. What do you want to do there?

**Comment by SpiritBob** [comment]
It's not anymore used only for that. As of ASP.NET Core 3.03/.1, it's the recommended and earliest place in a ASP.NET Core application to inject services. That means you can pretty much do both pipe-line configuration, as well as service configurations, which require already registered services.

The only services that are automatically injected **before** that point, are the main services, like IConfiguration, IWebHosting etc. Trying to inject anything else in the main method will yield in an exception.

**Comment by SpiritBob** [comment]
@leastprivilege if we want to implement the simplest of things, like logging certain events, using `ILogger,` we'll need to fetch the builder inside the Configure method.

**Comment by hugoqribeiro** [comment]
@SpiritBob - what you say about IConfiguration, etc is true before ConfigureServices(), not Configure().

**Comment by leastprivilege** [comment]
OK - I took the time to read through the announcement and attached issue/discussion.

Indeed things have changed in 3.0 that I was not aware of. 

You used to be able to inject an ILogger in the startup ctor to do logging during container composition. This feature is gone in 3.0 and Microsoft's official statement is, that logging is only supported after the DI containter is built - the earliest in `Configure`.

I personally don't like - nor do many others - but it is the way it is. There are several workarounds that include building your own logger or using Serilog (that's what I would do) to overcome this limitation.

This **DOES NOT** mean that you can fix this problem by moving your code from `ConfigureServices` to `Configure` like @SpiritBob tries to do. The builder is never put into the DI container, it is just a convenience helper around the raw DI config methods.

**Comment by SpiritBob** [comment]
@leastprivilege this was only the icing on the cake. If we ever implement our own Services, and/or other services, different than ILogger, would that mean that we'd have to add tons of more **workarounds**, if any, similar to the already proposed solutions (not really solutions) to having an ILogger/service instance available in the `ConfigureServices` method? Every workaround I've seen is an anti-pattern, especially ones involving calling `services.BuildServiceProvider();` and the reason why Microsoft made it like so - to avoid duplicate instancing of objects.

The least harmful way in ILogger's  case to solving the problem, is something along the lines of

```
        private ILogger GetEarlyInitializationLogger()
        {
            // Mimic behavior of original logger , to be used only in ConfigureServices,
            // as we have no control over DI loggers.
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                builder.AddConfiguration(_configuration.GetSection("Logging"));
                builder.AddConsole();
                builder.AddDebug();
                builder.AddEventSourceLogger();
            });

            return loggerFactory.CreateLogger("Initialization");
        }
```

Or, we can use a 3rd-party library, but why all this hassle, just to circumvent the new pattern made by Microsoft?

**Comment by leastprivilege** [comment]
You are asking the wrong person. sorry.

**Comment by leastprivilege** [comment]
The typical approach is that you simply add your own services to DI in `ConfigureServices` - once they get instantiated the container is done building and you have access to the other services.

But this is now out of scope for an IdentityServer issue. good luck.

**Comment by SpiritBob** [comment]
@leastprivilege I have access to the other services once the containe is done building, but not to IdentityServer's builder. That's the issue! If I'm to duplicate the builder for every service I want the only solution, doesn't this warrant some form of review?

**Comment by leastprivilege** [comment]
I think you might have a fundamental misunderstanding how that system works. Have a look at our builder source code - maybe that helps. In addition you can open an issue with Microsoft for the DI system.
